### PR TITLE
"Best practices"

### PR DESCRIPTION
**Pull request by magley**
Nothing special.

### CODE REVIEW

File: back/sssscs/.gitignore [hunk 1]
```diff
@@ -1,3 +1,7 @@
+# Keys and certs
+
+keys/*.key
+certs/*.crt
```
Comment (swopl) [inline back/sssscs/.gitignore]: Duplicate of lines 12-14 of this file

File: back/sssscs/build.gradle [hunk 2]
```diff
@@ -19,6 +20,7 @@ repositories {
 }
 
 dependencies {
+	implementation 'org.yaml:snakeyaml:2.0'
 	implementation "org.springframework.boot:spring-boot-starter-security"
```
Comment (swopl) [inline back/sssscs/build.gradle:24]: Where is this snakeYaml dep used?

File: back/sssscs/src/main/java/com/ib/user/UserService.java [hunk 3]
```diff
@@ -96,4 +85,24 @@ public Boolean isBlocked(User user) {
...
+		LocalDateTime lastTime = user.getLastTimeOf2FA();
+		
+		if (lastTime == null) {
+			return true;
+		}
+		
+		LocalDateTime now = LocalDateTime.now();
+		
+		long minutesPassed = ChronoUnit.MINUTES.between(lastTime, now);
+		return Math.abs(minutesPassed) > 5;
```
Comment (swopl) [inline back/sssscs/src/main/java/com/ib/user/UserService.java]: Why the Math.abs?

File: back/sssscs/src/main/java/com/ib/certificate/CertificateController.java [hunk 4]
```diff
@@ -64,7 +64,11 @@ private CertificateRequest requestFromCreateDto(CertificateRequestCreateDto dto)
 			}
 			req.setValidTo(desired);
 		} else {
-			req.setValidTo(LocalDateTime.now().plusMonths(2));
+			LocalDateTime desired = LocalDateTime.now().plusMonths(2);
+			if (desired.isAfter(req.getParent().getValidTo())) {
+				desired = req.getParent().getValidTo().minusDays(1);
+			}
+			req.setValidTo(desired);
```
Comment (swopl) [inline back/sssscs/src/main/java/com/ib/certificate/CertificateController.java]: This is duplicated code from other if. All branches of if set valid to desired time and bound it from top if parent exists. Extract common to outside of if, or into method

**Review by swopl — CHANGES_REQUESTED** [review]

File: back/sssscs/.gitignore (hunk 1 above)
Comment (magley) [reply back/sssscs/.gitignore]: [b9bff3a](https://github.com/magley/sssscs/pull/26/commits/b9bff3a1827ebae621a0629ac33f740045a866a0)

File: back/sssscs/build.gradle (hunk 2 above)
Comment (magley) [reply back/sssscs/build.gradle:24]: I can't remember and I can't be bothered to run the analysis again, but it's one of the libraries we directly include in our project. By default, that library depends on an older version of snakeyaml with [this](https://nvd.nist.gov/vuln/detail/CVE-2022-1471) vulnerability.

File: back/sssscs/src/main/java/com/ib/user/UserService.java (hunk 3 above)
Comment (magley) [reply back/sssscs/src/main/java/com/ib/user/UserService.java]: [70aab25](https://github.com/magley/sssscs/pull/26/commits/70aab2543bbe1213b6b7bf18365887faff7c2bb7)

File: back/sssscs/build.gradle (hunk 2 above)
Comment (swopl) [reply back/sssscs/build.gradle:24]: Ok, that's a pretty bad vuln

File: back/sssscs/src/main/java/com/ib/certificate/CertificateController.java (hunk 4 above)
Comment (magley) [reply back/sssscs/src/main/java/com/ib/certificate/CertificateController.java]: [7f1ca6f](https://github.com/magley/sssscs/pull/26/commits/7f1ca6f11edb7e307eeb633306737f000fb8367b)

File: back/sssscs/src/main/java/com/ib/certificate/CertificateController.java [hunk 5]
```diff
@@ -42,6 +42,20 @@ public class CertificateController {
 	private ModelMapper modelMapper;
 	@Autowired
 	private IAuthenticationFacade auth;
+	
+	private void setRequestValidTo(CertificateRequest req, long months) {
+		long years = months / 12;
+		months = months % 12;
+		
+		LocalDateTime desired = LocalDateTime.now().plusYears(years).plusMonths(months);
```
Comment (swopl) [inline back/sssscs/src/main/java/com/ib/certificate/CertificateController.java]: Does plusMonths(12) not work? Do you need to extract years?

File: back/sssscs/src/main/java/com/ib/certificate/CertificateController.java [hunk 6]
```diff
@@ -42,6 +42,20 @@ public class CertificateController {
 	private ModelMapper modelMapper;
 	@Autowired
 	private IAuthenticationFacade auth;
+	
+	private void setRequestValidTo(CertificateRequest req, long months) {
+		long years = months / 12;
```
Comment (swopl) [inline back/sssscs/src/main/java/com/ib/certificate/CertificateController.java]: Method name is valid to, but the method sets the length it is validFor. (validTo months -> validFor months)

**Review by swopl — CHANGES_REQUESTED** [review]

File: back/sssscs/src/main/java/com/ib/certificate/CertificateController.java (hunk 6 above)
Comment (magley) [reply back/sssscs/src/main/java/com/ib/certificate/CertificateController.java]: [270b5f9](https://github.com/magley/sssscs/pull/26/commits/270b5f9e7a41b34ed67171c61b4cccf8eb31d697)

File: back/sssscs/src/main/java/com/ib/certificate/CertificateController.java (hunk 5 above)
Comment (magley) [reply back/sssscs/src/main/java/com/ib/certificate/CertificateController.java]: [02123bf](https://github.com/magley/sssscs/pull/26/commits/02123bfbc58995304797cd059aa3353438c9fb79)

**Review by swopl — APPROVED** [review]

**Pull request merged**